crowbar_framework/app/models/service_object.rb

```diff
@@ -176,15 +176,22 @@ def set_to_applying(nodes, inst, pre_cached_nodes)
     end
   end
 
+  def restore_node_to_ready(node)
+    node.crowbar['state'] = 'ready'
+    node.crowbar['state_owner'] = ""
+    node.save
+  end
+
   def restore_to_ready(nodes)
     with_lock "BA-LOCK" do
       nodes.each do |node_name|
         node = NodeObject.find_node_by_name(node_name)
-        next if node.nil?
+        # Nodes with 'crowbar-upgrade' state need to stay in that state
+        # even after applying relevant roles. They could be brought back to
+        # being ready only by explicit user's action.
+        next if node.nil? || node.crowbar['state'] == 'crowbar-upgrade'
```

Comment:
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping. (https://github.com/bbatsov/ruby-style-guide#consistent-string-literals)